addons/account/models/account_move.py

```diff
@@ -1845,41 +1915,16 @@ def _check_unique_sequence_number(self):
 
     @api.constrains('ref', 'move_type', 'partner_id', 'journal_id', 'invoice_date', 'state')
     def _check_duplicate_supplier_reference(self):
-        moves = self.filtered(lambda move: move.state == 'posted' and move.is_purchase_document() and move.ref)
-        if not moves:
-            return
-
-        self.env["account.move"].flush([
-            "ref", "move_type", "invoice_date", "journal_id",
-            "company_id", "partner_id", "commercial_partner_id",
-        ])
-        self.env["account.journal"].flush(["company_id"])
-        self.env["res.partner"].flush(["commercial_partner_id"])
-
-        # /!\ Computed stored fields are not yet inside the database.
-        self._cr.execute('''
-            SELECT move2.id
-            FROM account_move move
-            JOIN account_journal journal ON journal.id = move.journal_id
-            JOIN res_partner partner ON partner.id = move.partner_id
-            INNER JOIN account_move move2 ON
-                move2.ref = move.ref
-                AND move2.company_id = journal.company_id
-                AND move2.commercial_partner_id = partner.commercial_partner_id
-                AND move2.move_type = move.move_type
-                AND (move.invoice_date is NULL OR move2.invoice_date = move.invoice_date)
-                AND move2.id != move.id
-            WHERE move.id IN %s
-        ''', [tuple(moves.ids)])
-        duplicated_moves = self.browse([r[0] for r in self._cr.fetchall()])
-        if duplicated_moves:
-            raise ValidationError(_('Duplicated vendor reference detected. You probably encoded twice the same vendor bill/credit note:\n%s') % "\n".join(
-                duplicated_moves.mapped(lambda m: "%(partner)s - %(ref)s - %(date)s" % {
-                    'ref': m.ref,
-                    'partner': m.partner_id.display_name,
-                    'date': format_date(self.env, m.invoice_date),
-                })
-            ))
+        """ Assert the move which is about to be posted isn't a duplicated move from another posted entry"""
+        move_to_duplicate_moves = self.filtered(lambda m: m.state == 'posted')._fetch_duplicate_supplier_reference(only_posted=True)
+        if move_to_duplicate_moves:
+            raise ValidationError(
+                _('Duplicated vendor reference detected. You probably encoded twice the same vendor bill/credit note:\n%s')
+                % "\n".join(
+                    "\n".join(duplicates.mapped(lambda m: f"  {m.partner_id.display_name} - {m.ref} - {format_date(self.env, m.invoice_date)}"))
+                    for duplicates in move_to_duplicate_moves.values()
```

Comment:
Won't this display twice the same values? Don't you need to look at `move_to_duplicate_moves.keys()` too?

Reply:
No it didn't but i was just me being dumb: it's ugly.